### DISCUSSION TITLE
Blocky with redis

### DISCUSSION THREAD

**Discussion by larivierec** [0xERR0R/blocky#557] [Q&A] [answered]
Good morning,

Yesterday I installed 0.19 of blocky in my cluster, however, I noticed that grafana blocky chart had really terrible average response times > 1s.

I then reverted back to 0.18 and also restarted redis to be sure and these response times were back to normal ~50ms. 

My question is should I have restarted redis when upgrading to 0.19? Would this have perhaps helped performance? Note I run in average of 2-5 replicas of blocky.

I could also create an issue if that could help. 

![image](https://user-images.githubusercontent.com/3633214/172619945-d686d17d-54ac-425f-8454-f8f021a32969.png)

I had to enable the filtering option on `AAAA` because the bootstrap was trying to resolve ipv6 and was timing out/failing for some of my file downloads.

I seem to get some broke pipes in the blocky logs
`[2022-06-08 13:41:23] ERROR can't write message: write tcp 172.16.134.239:853->192.168.10.204:45980: write: broken pipe`

these are the only things I have changed and for the time being have reverted.
https://github.com/larivierec/home-cluster/commit/05d40779411bf9aec5569f7f723cca72e63636d7

**Comment by ThinkChaos** [comment, marked as answer]
I don't think restarting Redis would do much, but I'm not very familiar with that code.

Any chance you can collect trace log for a query that takes a long time so we can narrow this down?

For the bootstrap issue, I think it's not the resolving that's problematic (since resolving `AAAA` over IPv4 is not an issue), but the connecting to the file host over HTTP.
The code uses a random IP from what was resolved for the host. And, unless you filtered `A` or `AAAA`, it resolves both. Meaning by default blocky will sometimes try to connect via IPv6 to the file host.
Is IPv6 connectivity not working in your blocky host?

The same behavior is used for any DoH upstream. That code will retry the query twice (3 tries total), so it might (randomly) start with IPv6, timeout/fail, retry and choose IPv4 at some point. Basically wasting time trying IPv6.
This shouldn't happen if you filtered `AAAA` queries though since there'd be no IPv6 to randomly choose. If I understand correctly you had those filtered when you measured those times; is that right?

The broken pipe error makes it sound like the client is disconnecting before blocky responds.
Do your clients have a shorter timeout than blocky?

One thing relevant here is that current `UpstreamResolver` code uses the configured timeout as the timeout for a single try.
So each request from a blocky client can take as much as 3x the configured timeout + 100ms of delay in between each try (so 2 x 100ms).
You might want to lower your timeout config, but that doesn't solve your long average response time, unless you're getting a lot of timeouts.
@0xERR0R would you be open to changing how the timeout configuration works?  
Maybe we should open an issue to discuss that, but I think expanding the config to allow changing the number of tries, and their delay, would help make this behavior clear to the user.

**Reply by 0xERR0R** [reply]
> @0xERR0R would you be open to changing how the timeout configuration works? Maybe we should open an issue to discuss that, but I think expanding the config to allow changing the number of tries, and their delay, would help make this behavior clear to the user.

If we change the behaviour, it will "break" existing installations. Maybe it should be better described in the docs, with more examples?

**Reply by larivierec** [reply]
Hey there @ThinkChaos sorry for the delay in my reply, yours was quite thorough and I wanted to make sure that I answer everything properly.

> Any chance you can collect trace log for a query that takes a long time so we can narrow this down?

Which option would you wish I enable for this?
Is my current configuration enough?

## Topology clarification

My network topology isn't so complex, I have most of my network separated with vlans and everything can hit my DNS server at 192.168.42.4 which is blocky. If it's my own domain, blocky forwards the request to a different DNS server called k8s-gateway which also resides in the cluster and only resolves those specific DNS names, it is basically CoreDNS but with extra features.

Blocky == master
k8s-gateway == slave 

> For the bootstrap issue, I think it's not the resolving that's problematic (since resolving `AAAA` over IPv4 is not an issue), but the connecting to the file host over HTTP. The code uses a random IP from what was resolved for the host. And, unless you filtered `A` or `AAAA`, it resolves both. Meaning by default blocky will sometimes try to connect via IPv6 to the file host. Is IPv6 connectivity not working in your blocky host?

This is weird because while it starts, it seems fine with 0.18 without the filtering `AAAA` option (see output below)
```
[2022-06-09 20:51:14]  INFO list_cache: starting download link=https://www.github.developerdan.com/hosts/lists/ads-and-tracking-extended.txt
[2022-06-09 20:51:14]  INFO list_cache: starting download link=https://pgl.yoyo.org/adservers/serverlist.php?hostformat=hosts&showintro=0&mimetype=plaintext
[2022-06-09 20:51:14]  INFO list_cache: starting download link=https://raw.githubusercontent.com/jdlingyu/ad-wars/master/hosts
[2022-06-09 20:51:14]  INFO list_cache: starting download link=https://raw.githubusercontent.com/crazy-max/WindowsSpyBlocker/master/data/hosts/spy.txt
[2022-06-09 20:51:14]  INFO list_cache: starting download link=https://raw.githubusercontent.com/bigdargon/hostsVN/master/hosts
[2022-06-09 20:51:14]  INFO list_cache: starting download link=https://raw.githubusercontent.com/FadeMind/hosts.extras/master/add.2o7Net/hosts
[2022-06-09 20:51:14]  INFO list_cache: starting download link=https://adaway.org/hosts.txt
[2022-06-09 20:51:14]  INFO list_cache: starting download link=https://v.firebog.net/hosts/lists.php?type=nocross
[2022-06-09 20:51:14]  INFO list_cache: starting download link=https://raw.githubusercontent.com/StevenBlack/hosts/master/hosts
[2022-06-09 20:51:14]  INFO list_cache: starting download link=https://raw.githubusercontent.com/anudeepND/blacklist/master/adservers.txt
```

When updating to 0.19, It does not download all of them and I get a diff in blocked domains. (this is without the filtering option)
~380k in 0.19 vs 550k in 0.18.

If I use the filtering option with `AAAA` then I receive the same amount of blocks than 0.18

IPv6 is not disabled on my network however, it is not encouraged as 100% of my networking is done using IPv4.

> The broken pipe error makes it sound like the client is disconnecting before blocky responds. Do your clients have a shorter timeout than blocky?

The broken pipe only seems to happen with specific devices... it could be the access points or etc... but if I revert to 0.18 everythitng is fine and dandy. Also, if I do a request it is REALLY long on my devices > 3-5s it's as if it gets stuck.

> One thing relevant here is that current `UpstreamResolver` code uses the configured timeout as the timeout for a single try. So each request from a blocky client can take as much as 3x the configured timeout + 100ms of delay in between each try (so 2 x 100ms). You might want to lower your timeout config, but that doesn't solve your long average response time, unless you're getting a lot of timeouts. @0xERR0R would you be open to changing how the timeout configuration works? Maybe we should open an issue to discuss that, but I think expanding the config to allow changing the number of tries, and their delay, would help make this behavior clear to the user.

I could also try that probably put it to something more aggressive. Is this new or was always like this?

Thanks again for your reply, if it could help, I could spawn a different deployment with a few instances and throw some traffic on it :)
Just let me know what you'd like me to collect.

Note: blocky is awesome, about time we have a stateless dns server :1st_place_medal:

**Reply by ThinkChaos** [reply]
> Which option would you wish I enable for this?

The most important is `logLevel: trace` so we can get as much info as possible.  
And doing it on a separate instance where you just run queries manually would be awesome to reduce the number of logs since blocky will be quite verbose!

> When updating to 0.19, It does not download all of them and I get a diff in blocked domains.

Blocky doesn't fail when a list download/parse errors, so this is compatible with the theory of IPv6 not working.

Can you try downloading a list that blocky warns about in its logs with `curl --verbose --ipv6`?  
We'll know if there's IPv6 issues outside of blocky's control for sure if that fails.
The blocky log would still be interesting just to confirm it also trips up on IPv6.

> The broken pipe only seems to happen with specific devices... it could be the access points or etc... but if I revert to 0.18 everythitng is fine and dandy. Also, if I do a request it is REALLY long on my devices > 3-5s it's as if it gets stuck.

Does this happen even if you filter IPv6 queries?  
Also have the same question for the latency. All the issues seem like they could be caused by that.

> I could also try that probably put it to something more aggressive. Is this new or was always like this?

Not sure about "always", but that hasn't changed recently.

**Reply by larivierec** [reply]
Hello and sorry for the delay, like many, i've been quite busy the last few days.
I spawned another instance of blocky in the cluster on a different IP here is the full file, you'll see the configuration in the `config:` section.
https://github.com/larivierec/home-cluster/blob/main/cluster/apps/networking/test/blocky-upgrade/helm-release.yaml

I removed the ingress for api and dependency to redis + removed local dns over tls and https (most of my devices use 53 udp/tcp).

I tried to do the curl as you asked as well. It's producing the same output as blocky.
```bash
(⎈ |k3s:default) ~/personal_code/github/home-cluster (main) > curl --verbose --ipv6 https://www.github.developerdan.com/hosts/lists/ads-and-tracking-extended.txt
*   Trying 2606:50c0:8000::153:443...
* Immediate connect fail for 2606:50c0:8000::153: Network is unreachable
*   Trying 2606:50c0:8003::153:443...
* Immediate connect fail for 2606:50c0:8003::153: Network is unreachable
*   Trying 2606:50c0:8002::153:443...
* Immediate connect fail for 2606:50c0:8002::153: Network is unreachable
*   Trying 2606:50c0:8001::153:443...
* Immediate connect fail for 2606:50c0:8001::153: Network is unreachable
* Closing connection 0
curl: (7) Couldn't connect to server
```

Also, a weird log I noticed when using DoH of cloudflare -> 
```
[2022-06-19 15:40:25] DEBUG parallel_best_resolver: resolution failed from resolver, cause: can't resolve request via upstream server https://104.16.249.249:443/dns-query: http return code should be 200, but received 403
```
Note: my cluster is not configured as "Dual-Stack" only IPv4 however, this was working fine before. I guess it was resolving everything as IPv4.

I have a log output of the bootstrap and the list size is definitely not the same.
~ 550k + (blocky 0.18)
~ 175k (blocky 0.19) because of the network unreacheable.

The latter baffles me :|

If there's more I could supply let me know. :)
See attached file for bootstrap
[bootstrap-dual-stack.txt](https://github.com/0xERR0R/blocky/files/8935910/bootstrap-dual-stack.txt)

**Reply by ThinkChaos** [reply]
> I tried to do the curl as you asked as well. It's producing the same output as blocky.

Since this is a network issue outside of blocky's controle, I think it's hard to consider this a blocky bug.  
Easiest fix should be for you to filter `AAAA`. Maybe adding a separate option to tell blocky to not use IPv6, without filtering `AAAA` could be added. Would that be useful to you?

Besides that I had worked on adding a very basic way for blocky to remember "good" vs "bad" upstream IPs (basically use the IP that last worked until there's an error, at which point try the next one). In the end I had removed that because it was too much complexity for little gains. I tried again and have a simpler implementation which might be worth integrating but that's up to @0xERR0R. See https://github.com/0xERR0R/blocky/pull/571.  
This could alleviate issues like yours, but is not a fix either.

**Reply by 0xERR0R** [reply]
Maybe we should improve the network detection? According to following log entries:

```
[2022-06-19 15:40:21] TRACE bootstrap: dialing pgl.yoyo.org addr=pgl.yoyo.org:443 ip=213.230.210.231 network=tcp
[2022-06-19 15:40:21] DEBUG upstream_resolver: received response from upstream answer=AAAA (2606:50c0:8001::154), AAAA (2606:50c0:8003::154), AAAA (2606:50c0:8002::154), AAAA (2606:50c0:8000::154) net=tcp-tls protocol=TCP response_time_ms=28 return_code=NOERROR upstream=tcp-tls:one.one.one.one upstream_ip=1.1.1.1
```

Blocky asks the bootstrap DNS (tcp-tls:one.one.one.one) for AAAA records. I assume, this code fragment from bootstrap.go:

```
switch {
			case strings.HasSuffix(network, "4") || filteredQTypes.Contains(dns.Type(dns.TypeAAAA)):
				qTypes = []dns.Type{dns.Type(dns.TypeA)}
			case strings.HasSuffix(network, "6") || filteredQTypes.Contains(dns.Type(dns.TypeA)):
				qTypes = []dns.Type{dns.Type(dns.TypeAAAA)}
			default:
				qTypes = v4v6QTypes
			}
```

is too optimistic. What do you think?

**Reply by ThinkChaos** [reply]
I think disabling IPv6 when we get errors is a bit of an overreaction and don't have better ideas for auto-detection (maybe there's proper ways to do that but I didn't find any in my quick searches). The mentioned MR is a half solution: we don't fully disable IPv6, just try to avoid bad IPs (whatever their version).  
IMO it's up to the admin to say whether they want it or not, so I think a second option for "don't use IPv6 but allow `AAAA` lookups" would be perfect for this. It could be used in the code you quoted.

**Reply by 0xERR0R** [reply]
Since the IP of the bootstrap is known, maybe we should assume, that we can always do A query if it is IPv4 and AAAA for IPv6?

**Reply by ThinkChaos** [reply]
So this would be the behavior?

|⬇️bootstrap IPs, ➡️Connect using|v4|v6|
|---|---|---|
|None|✔️|✔️|
|v4|✔️| |
|v6| |✔️|
|Both|✔️|✔️|

I think it's sensible default behavior but also think adding an option to explicit control IPv6 usage would be beneficial since more explicit.

**Reply by 0xERR0R** [reply]
Yes, this could work. We could also provide a new config option "bootstrapNetwork" (or something more meaningful) as enum with ipv4, upv6 or both as values. Default is both.

**Reply by ThinkChaos** [reply]
I just realized GitHub doesn't notify a discussion when it has been mentioned, so doing it manually: PR https://github.com/0xERR0R/blocky/pull/589 is linked.

**Reply by larivierec** [reply]
Thanks again for the feature, working like a charm 😃
